### PR TITLE
Add email notifications for organiser to appointments

### LINKED ISSUE #3630: Create Notification for Organiser when booking is confirmed / cancelled
Currently, the appointment owner ( = organiser) has no way to know which slots have been booked unless they check their calendar. It would be nice to have an event notification for each confirmed booking. 

Thinking a bit further, this could also be useful in case a booking user decides to update their PARTSTAT from Calendar et al directly. As we don't offer a "Cancel booking link" at the moment, this is not unlikely to happen.

This will allow the organiser to be more cognisant of their appointments.

Example

### PR DESCRIPTION
**Pull request by miaulalala**
Fixes #3630

Signed-off-by: Anna Larch <anna@nextcloud.com>

### CODE REVIEW

File: lib/Notification/Notifier.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+ */
+
+namespace OCA\Calendar\Notification;
+
+use OCP\IL10N;
+use OCP\IURLGenerator;
+use OCP\L10N\IFactory;
+use OCP\Notification\INotification;
+
+class Notifier implements \OCP\Notification\INotifier {
```
Comment (ChristophWurst) [inline lib/Notification/Notifier.php:33]: import

File: lib/Notification/Notifier.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	 * Human-readable name describing the notifier
+	 * @return string
+	 */
+	public function getName(): string {
+		return $this->factory->get('calendar')->t('Calendar');
+	}
+
+
+	public function prepare(INotification $notification, string $languageCode): INotification {
+		if ($notification->getApp() !== 'calendar') {
```
Comment (ChristophWurst) [inline lib/Notification/Notifier.php:58]: Application::APP_ID

**Review by ChristophWurst — CHANGES_REQUESTED** [review]

File: lib/AppInfo/Application.php [hunk 3]
```diff
@@ -56,6 +56,9 @@ public function register(IRegistrationContext $context): void {
 		}
 
 		$context->registerEventListener(UserDeletedEvent::class, UserDeletedListener::class);
+
+		$manager = \OC::$server->get(\OCP\Notification\IManager::class);
+		$manager->registerNotifierService(\OCA\Calendar\Notification\Notifier::class);
```
Comment (tcitworld) [inline lib/AppInfo/Application.php:61]: **nitpick:** You can use the `registerNotifierService` directly from the `$context` parameter.

File: lib/Service/Appointments/MailService.php [hunk 4]
```diff
@@ -247,4 +254,99 @@ private function getSysEmail(): string {
...
+		}
+
+		$fromEmail = $user->getEMailAddress();
+		$fromName = $user->getDisplayName();
+
+		$sys = $this->getSysEmail();
+		$message = $this->mailer->createMessage()
+			->setFrom([$sys => $fromName])
+			->setTo([$user->getEMailAddress() => $booking->getDisplayName()])
+			->setReplyTo([$fromEmail => $fromName]);
```
Comment (tcitworld) [inline lib/Service/Appointments/MailService.php:273]: **question:** Why would I want to reply to myself?

**Review by tcitworld — DISMISSED** [review]

File: lib/Service/Appointments/MailService.php (hunk 4 above)
Comment (tcitworld) [reply lib/Service/Appointments/MailService.php]: Update tests as well

File: lib/AppInfo/Application.php [hunk 5]
```diff
@@ -56,6 +56,8 @@ public function register(IRegistrationContext $context): void {
 		}
 
 		$context->registerEventListener(UserDeletedEvent::class, UserDeletedListener::class);
+
+		$context->registerNotifierService(\OCA\Calendar\Notification\Notifier::class);
```
Comment (ChristophWurst) [inline lib/AppInfo/Application.php:60]: import

File: lib/Notification/Notifier.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+				]);
+				$placeholders = $replacements = [];
+				foreach ($notification->getRichSubjectParameters() as $placeholder => $parameter) {
+					$placeholders[] = '{' . $placeholder . '}';
+					$replacements[] = $parameter[$placeholder];
+				}
+				$notification->setParsedSubject(str_replace($placeholders, $replacements, $notification->getRichSubject()));
+
+				$messageParameters = $notification->getMessageParameters();
+				$notification->setRichMessage($l->t('{display_name} ({email}) has booked the appointment "{config_display_name}" on {date_time}.'), [
```
Comment (ChristophWurst) [inline lib/Notification/Notifier.php:86]:
```suggestion
				$notification->setRichMessage($l->t('{display_name} ({email}) booked the appointment "{config_display_name}" on {date_time}.'), [
```

:thinking:

File: lib/Service/Appointments/BookingService.php [hunk 7]
```diff
@@ -114,11 +114,20 @@ public function confirmBooking(Booking $booking, AppointmentConfig $config): Boo
...
+			$this->mailService->sendOrganizerBookingInformationEmail($booking, $config, $calendar);
 		} catch (ServiceException $e) {
-			$this->logger->info('Could not send booking information email after confirmation by user ' . $booking->getEmail(), ['exception' => $e]);
+			$this->logger->info('Could not send booking emails after confirmation from user ' . $booking->getEmail(), ['exception' => $e]);
 		}
+
+		try {
+			$this->mailService->sendOrganizerBookingInformationNotification($booking, $config);
+		} catch (\InvalidArgumentException $e) {
+			$this->logger->info('Could not send booking information notification after confirmation by user ' . $booking->getEmail(), ['exception' => $e]);
```
Comment (ChristophWurst) [inline lib/Service/Appointments/BookingService.php:128]: nit: make it a warning or error unless there is a good reason to silently ignore error here

File: lib/Service/Appointments/MailService.php [hunk 8]
```diff
@@ -247,4 +254,97 @@ private function getSysEmail(): string {
...
+
+		$template = $this->mailer->createEMailTemplate('calendar.confirmOrganizer');
+		$template->addHeader();
+
+		//Subject
+		$subject = $this->l10n->t('You have a new appointment booking "%s" from %s', [$config->getName(), $booking->getDisplayName()]);
+		$template->setSubject($subject);
+
+		// Heading
+		$summary = $this->l10n->t('Dear %s, %s (%s) has booked an appointment with you.', [$user->getDisplayName(), $booking->getDisplayName(), $booking->getEmail()]);
```
Comment (ChristophWurst) [inline lib/Service/Appointments/MailService.php:282]:
```suggestion
		$summary = $this->l10n->t('Dear %s, %s (%s) booked an appointment with you.', [$user->getDisplayName(), $booking->getDisplayName(), $booking->getEmail()]);
```

just a detail but again unsure about the correct grammar

File: lib/Service/Appointments/MailService.php [hunk 9]
```diff
@@ -247,4 +254,97 @@ private function getSysEmail(): string {
...
+		$sys = $this->getSysEmail();
+		$message = $this->mailer->createMessage()
+			->setFrom([$sys => $fromName])
+			->setTo([$user->getEMailAddress() => $booking->getDisplayName()]);
+
+
+		$template = $this->mailer->createEMailTemplate('calendar.confirmOrganizer');
+		$template->addHeader();
+
+		//Subject
```
Comment (ChristophWurst) [inline lib/Service/Appointments/MailService.php:277]:
```suggestion
		// Subject
```

File: lib/Service/Appointments/MailService.php [hunk 10]
```diff
@@ -247,4 +254,97 @@ private function getSysEmail(): string {
...
+			$failed = $this->mailer->send($message);
+			if (count($failed) > 0) {
+				$this->logger->warning('Mail delivery failed for some recipients.');
+				foreach ($failed as $fail) {
+					$this->logger->debug('Failed to deliver email to ' . $fail);
+				}
+				throw new ServiceException('Could not send mail for recipient(s) ' . implode(', ', $failed));
+			}
+		} catch (Exception $ex) {
+			$this->logger->error($ex->getMessage(), ['exception' => $ex]);
```
Comment (ChristophWurst) [inline lib/Service/Appointments/MailService.php:309]:
```suggestion
			$this->logger->error("Could not send appointment organizer email: " . $ex->getMessage(), ['exception' => $ex]);
```

File: tests/php/unit/Notification/NotifierTest.php [hunk 11]
```diff
@@ -0,0 +1,200 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * *
+ *  * Calendar App
```
Comment (ChristophWurst) [inline tests/php/unit/Notification/NotifierTest.php:7]:
```suggestion
 * Calendar App
```

File: tests/php/unit/Notification/NotifierTest.php [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+			->method('get')
+			->willReturn($this->l);
+
+		$this->notifier = new Notifier(
+			$this->factory,
+			$this->url
+		);
+	}
+
+	public function testPrepareWrongApp() {
```
Comment (ChristophWurst) [inline tests/php/unit/Notification/NotifierTest.php:67]: void for all test methods

File: tests/php/unit/Notification/NotifierTest.php [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+			->willReturn('calendar');
+		$notification->expects($this->once())
+			->method('getSubject')
+			->willReturn('wrong subject');
+
+		$this->notifier->prepare($notification, 'en');
+	}
+
+	public function testPrepare() {
+		/** @var \OCP\Notification\INotification|\PHPUnit\Framework\MockObject\MockObject $notification */
```
Comment (ChristophWurst) [inline tests/php/unit/Notification/NotifierTest.php:97]: import

File: lib/Notification/Notifier.php (hunk 6 above)
Comment (miaulalala) [reply lib/Notification/Notifier.php:86]: I think both is correct? But "has booked" kinda implies another action is neccessary for the booking to be complete, which isn't true in our case so I will change it to "booked"

File: lib/Service/Appointments/BookingService.php (hunk 7 above)
Comment (miaulalala) [reply lib/Service/Appointments/BookingService.php:128]: just using the same logic as above for the emails.

Comment (ChristophWurst) [reply lib/Service/Appointments/BookingService.php]: please give me a chance to see comments that do not get addressed. leave these github review comments unresolved.

Comment (ChristophWurst) [reply lib/Service/Appointments/BookingService.php]: it did get changed. alright.

**Pull request merged**